### PR TITLE
Adds exception hook from IR exceptions

### PR DESCRIPTION
**Pull request by aopincar**
IR exceptions are now handled by a dedicated hook.
Replace all exception with proper IRException subclasses and
remove all sys.exit calls.

relates to  #6

### CODE REVIEW

File: tools/kcli/kcli/conf.py [hunk 1]
```diff
@@ -1,12 +1,13 @@
 import ConfigParser
 import os
+from kcli.exceptions import IRFileNotFoundException
```
Comment (yfried) [inline tools/kcli/kcli/conf.py]: `from kcli import exceptions`
Please fix L1 as well.

also, different import group.

File: tools/kcli/kcli/conf.py [hunk 2]
```diff
@@ -24,18 +25,14 @@ def load_config_file():
...
-    print "ERROR: kcli conf file (%s) not found" % KCLI_CONF_FILE
-    sys.exit(1)
-
-config = load_config_file()
-
-
 
+    conf_file_paths = "\n".join([CWD_PATH, USER_PATH, SYSTEM_PATH])
+    raise IRFileNotFoundException(conf_file_paths,
+                                  "IR configuration doesn't found, please "
```
Comment (yfried) [inline tools/kcli/kcli/conf.py]: "IR configuration not found. Please "

File: tools/kcli/kcli/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+class IRException(Exception):
+    def __init__(self, msg):
+        self.msg = msg
+
+
+class IRKeyNotFoundException(IRException):
+    def __init__(self, key, dic):
+        super(self.__class__, self).__init__(
+            "Key \"%s\" not found in %s" % (key, dic))
```
Comment (yfried) [inline tools/kcli/kcli/exceptions.py]: `'Key "%s" not found in %s'`

File: tools/kcli/kcli/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+class IRException(Exception):
+    def __init__(self, msg):
+        self.msg = msg
```
Comment (yfried) [inline tools/kcli/kcli/exceptions.py]: use `message`

File: tools/kcli/kcli/exceptions.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+            "in 'key=value' form" % extra_var)
+
+
+class IRMissingAncestorException(IRException):
+    def __init__(self, key):
+        super(self.__class__, self).__init__(
+            "Please provide all ancestors of \"--%s\"" % key.replace("_", "-"))
+
+
+class IRMissingEnvironmentVariableExcption(IRException):
```
Comment (yfried) [inline tools/kcli/kcli/exceptions.py]: - s/Missing/Undefined/

If type is Undefined var, we only need the var name in message

File: tools/kcli/kcli/logger.py [hunk 6]
```diff
@@ -16,6 +20,28 @@
...
+
+def kcli_traceback_handler(log):
+    """Creates exception hook that sends IRException to log and other
+    exceptions to stdout (default excepthook)
+    :param log: logger to log trace
+    """
+
+    def my_excepthook(exc_type, exc_value, exc_traceback):
+        # sends full exception with trace to log
+        if not isinstance(exc_value, IRException):
```
Comment (yfried-redhat) [inline tools/kcli/kcli/logger.py]: consider using `issubclass`

File: tools/kcli/kcli/main.py [hunk 7]
```diff
@@ -8,16 +8,17 @@
 import yaml
 import configure
 
+from kcli import logger
 from kcli import conf
 from kcli.execute.execute import PLAYBOOKS
-from kcli import logger
 from kcli import parse
+from kcli.exceptions import *
```
Comment (yfried) [inline tools/kcli/kcli/main.py]: `from kcli import exceptions`

File: tools/kcli/kcli/main.py [hunk 8]
```diff
@@ -47,9 +48,7 @@ def dict_lookup(dic, key, *keys):
         LOG.debug("value has been found: \"%s\"" % value)
         return value
     except KeyError:
-        err_msg = "Key \"%s\" not found in %s" % (key, dic)
-        LOG.error(err_msg)
-        sys.exit(1)
+        raise IRKeyNotFoundException(key, dic)
```
Comment (yfried) [inline tools/kcli/kcli/main.py]: Try/except should focus only on L47

File: tools/kcli/kcli/conf.py [hunk 9]
```diff
@@ -24,18 +25,15 @@ def load_config_file():
...
-
-config = load_config_file()
-
-
 
+    conf_file_paths = "\n".join([CWD_PATH, USER_PATH, SYSTEM_PATH])
+    raise exceptions.IRFileNotFoundException(
+        conf_file_paths,
+        "IR configuration not found. "
+        "Please set it in one of the following paths:\n")
```
Comment (rhosqeauto) [inline tools/kcli/kcli/conf.py:33]: forgot to add paths

**Comment by rhosqeauto** [comment]
Merged a rebased version

**Pull request closed without merging**